cloud-sdk-tool-build-task/cloud-sdk-package.ts

```diff
@@ -0,0 +1,171 @@
+﻿// Copyright 2017 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import {getDefaultExecOptions} from 'common/exec-options';
+import * as http from 'http';
+import * as os from 'os';
+import * as path from 'path';
+import * as task from 'vsts-task-lib/task';
+import * as toolLib from 'vsts-task-tool-lib/tool';
+
+type ValidPlatform = 'win32'|'linux'|'darwin';
+type ValidArch = 'x64'|'x86'|'ia32';
+
+function isValidPlatform(platform: string): platform is ValidPlatform {
+  return platform === 'win32' || platform === 'linux' || platform === 'darwin';
+}
+
+function isValidArch(arch: string): arch is ValidArch {
+  return arch === 'x64' || arch === 'x86' || arch === 'ia32';
+}
+
+const cloudSdkId = 'google-cloud-sdk';
+const urlRoot = 'http://dl.google.com/dl/cloudsdk/channels/rapid';
+const downloadUrlRoot = `${urlRoot}/downloads`;
+const versionDoc = `${urlRoot}/components-2.json`;
+
+export class CloudSdkPackage {
+  readonly version: string;
+  private toolPath: string;
+
+  constructor(versionString: string, di: {toolLib: typeof toolLib}) {
+    this.version = di.toolLib.cleanVersion(versionString);
+    if (!(this.version && this.version.length > 0)) {
+      throw new Error(`Given version ${versionString} is not a valid version.`);
+    }
+    this.toolPath = di.toolLib.findLocalTool(cloudSdkId, this.version);
+  }
+
+  static async queryLatestVersion(di: {http: typeof http}): Promise<string> {
+    const message = await new Promise<http.IncomingMessage>(
+        (resolve,
+         reject) => { di.http.get(versionDoc, resolve).on('error', reject); });
+
+    const data = await new Promise<string>((resolve) => {
+      let rawData = '';
+      message.on('data', (chunk: string) => rawData += chunk);
+      message.on('end', () => resolve(rawData));
+    });
+    return JSON.parse(data)['version'] as string;
+  }
+
+  static async createPackage(versionSpec?: string, di = {http, task, toolLib}):
+      Promise<CloudSdkPackage> {
```

Comment:
What do you think about letting `createPackage` handle the `init` vs `acquire` logic? That way I think we don't even need the `isCached` function.

Reply:
I don't think this is the right place for that logic. I added a new function to keep that logic. In any case, I can't get rid of the `isCached` function.